syntax = "proto3";

import "google/protobuf/timestamp.proto";

package dfs_service;

service DFSService {

    // Add your service calls here

    // 1. REQUIRED (Parts 1 & 2): A method to store files on the server
    rpc StoreFile (stream FileContent) returns (FileName) {}

    // 2. REQUIRED (Parts 1 & 2): A method to fetch files from the server
    rpc FetchFile (FileRequest) returns (stream FileContent) {}

    // 3. REQUIRED (Parts 1 & 2): A method to list all files on the server
    rpc ListFiles (Empty) returns (FilesList) {}

    // 4. REQUIRED (Parts 1 & 2): A method to get the status of a file on the server
    rpc GetFileStatus (FileRequest) returns (FileStatus) {}

    // 5. REQUIRED (Part 2 only): A method to request a write lock from the server
    rpc GetWriteLock (FileInfo) returns (Result) {}

    // 6. REQUIRED (Part 2 only): A method named CallbackList to handle asynchronous file listing requests
    //                            from a client. This method should return a listing of files along with their
    //                            attribute information. The expected attribute information should include name,
    //                            size, modified time, and creation time.
    rpc CallbackList (FileInfo) returns (FilesList) {}

    // 7. REQUIRED (Part 2 only): A method to delete a file from the server
    rpc DeleteFile (FileRequest) returns (Empty) {}

    // 8. Any other methods you deem necessary to complete the tasks of this assignment


}

// Add your message types here

message FileInfo {
    string name = 1;
}

message Result {
    string status = 1;
}

message FileRequest {
    string filename = 1;
    uint32 checksum = 2;
}

message FileContent {
    bytes data = 1;
    uint32 size = 2;
}

message FileName {
    string filename = 1;
}

message Empty {

}

message FileStat {
    string filename = 1;
    google.protobuf.Timestamp modified = 2;
}

message FilesList {
    repeated FileStatus file = 1;
}

message FileStatus {
    string filename = 1;
    uint64 size = 2;
    google.protobuf.Timestamp modified = 3;
    google.protobuf.Timestamp created = 4;
}